feature: dotnet core 3.1 (#516)
GoogleContainerTools/distroless: experimental/dotnet_core_runtime/README.md

# Microsoft .NET Core Distroless Container Image

This project defines a minimal Linux with [.NET Core](https://github.com/dotnet/core).

> Specifically, the image contains everything described in the [../../cc/README.md](../../cc/README.md), and includes following prerequisites.

Reviewer: ```suggestion
Specifically, the image contains everything described in the [base image](../../cc/README.md), plus .NET Core and necessary dependencies.
```

> ## NET Core 3.1 Runtime and its Dependencies

Reviewer: ```suggestion
## .NET Core 3.1 Runtime and its Dependencies
```

- [3.1/runtime-deps/buster-slim/amd64/Dockerfile](https://github.com/dotnet/dotnet-docker/blob/master/3.1/runtime-deps/buster-slim/amd64/Dockerfile)
- [3.1/runtime/buster-slim/amd64/Dockerfile](https://github.com/dotnet/dotnet-docker/blob/master/3.1/runtime/buster-slim/amd64/Dockerfile)

> ```text

Reviewer: Since this is a long list, this can easily go out of sync or change going forward. Library versions will also vary depending on whether `-debian9`, `-debian10`, `-debian11`, etc. I think we can instead put a link to the Bazel file instead of explicitly listing debs here.

Author: I followed the content from existing [README.md](https://github.com/GoogleContainerTools/distroless/blob/master/experimental/dotnet/README.md#image-contents).

Author: Thank you. PR updated.

        packages["libidn2-0"],
        packages["libk5crypto3"],
        packages["libkrb5-3"],
        packages["libldap-2.4-2"],
        packages["libldap-common"],
        packages["libsasl2-2"],
        packages["libnghttp2-14"],
        packages["libpsl5"],
        packages["librtmp1"],
        packages["libssh2-1"],
        packages["libkeyutils1"],
        packages["libkrb5support0"],
        packages["libunistring0"],
        packages["libgnutls30"],
        packages["libgmp10"],
        packages["libhogweed4"],
        packages["libidn11"],
        packages["libnettle6"],
        packages["libp11-kit0"],
        packages["libffi6"],
        packages["libtasn1-6"],
        packages["libsasl2-modules-db"],
        packages["libdb5.3"],
        packages["libgcrypt20"],
        packages["libgpg-error0"],
        packages["libacl1"],
        packages["libattr1"],
        packages["libselinux1"],
        packages["libpcre3"],
        packages["libbz2-1.0"],
        packages["liblzma5"],
    ],
```

These dependencies are from the .NET Core `runtime` and `runtime-deps` that also install `curl`.

```bash
apt-get install apt-rdepends
apt-rdepends curl
```

> 
> Bazel does _not_ support dotnet build rules yet. ~See [#39](https://github.com/bazelbuild/rules_dotnet/issues/39).~ Unlike other distroless rules in this repository, users will need to compile and generate your .NET binary outside of Bazel manually. Instructions to generate the default binary files is described below.
> 
> ## Usage
> 
> The `ENTRYPOINT` of this image is left unset, so this image expects users to supply a the full `dotnet` and related path to the generated `*.dll` file in the `CMD`.

Reviewer: I think this is unnecessary.

Note the target audience of this repo is the consumers of the published Distroless images on gcr.io, not Bazel users who build Distroless images from Distroless source. We never supported Bazel users and will not.

```suggestion
```

### Examples

See [../../examples/dotnet_core/Dockerfile](../../examples/dotnet_core/Dockerfile).

> 
> See [../../local_build_dotnet_core_example.sh](../../local_build_dotnet_core_example.sh) script.

Reviewer: ```suggestion
```

### Prerequisites

We need to build a binary out-of-band since the binary build happens outside of Bazel.

It requires a local installation of .NET Core runtime.

> See [.NET Core 3.1.4 - May 12, 2020](https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1.4/3.1.4.md) release.

Reviewer: Likewise, I don't think this is necessary. It's enough to have a working `Dockerfile` with the published images.

Author: I agree to some extent. But, how should we explain to newbie folks like myself who are getting started in contribution when we are targeting a new non-existing container image from the container registry? For example, this PR introduces three separate dotnet core container images for the first time.

In addition, this section is based mostly from existing [README.md](https://github.com/GoogleContainerTools/distroless/blob/master/experimental/dotnet/README.md#prerequsite).

Reviewer: Alright, that's fair then. However, I'd like to move the relevant part to a separate doc for internal devs and contributors. To reduce unnecessary back-and-forth, would you mind if I directly modify this PR regarding this part?

And as I said, the README was added a very long time ago, and we have been making clear the goal of this repo recently, as we've seen a growing number of people misunderstanding the intent of the repo.

Author: Yes, please modify this PR in regards to internal development and contributors. Thanks for your help!

> 
> ### Build Locally with Bazel or Manually
> 
> See [../../local_build_dotnet_core_example.sh](../../local_build_dotnet_core_example.sh) script.

Reviewer: ```suggestion
```
